File: src/Frame.py
"""Class to store the essential information needed for a UDP frame
"""
from FrameType import FrameType


class Frame:
    """Class to represent a UDP frame.

    Contains the essential components of a UDP frame for communication between stations
    as well methods to convert between string format, byte format and to initiliase an
    instance from a string representation.
    Attributes:
        origin: the name of the origin node for the frame
        dest: the name of the node which is the final destination of the frame
        src: a list containing the names of the nodes the frame has passed through,
        ordered chronologically
        seqno: the sequence number of the frame applied by the origin
        time: the time, in minutes after midnight that the previous sender can arrive
        at the node it is sent to
        type: an enum representing the frames type
    """

    origin: str = ""
    dest: str = ""
    src: list
    seqno: int = -1
    time: int = -1
    type: FrameType

    def __init__(self, origin="", dest="", src=[], seqno=-1, time=-1, type=-1):
        """Inits Frame object
        """
        self.origin = origin
        self.dest = dest
        self.src = src
        self.seqno = seqno
        self.time = time
        self.type = type

    def to_string(self):
        """Returns string representation of frame object for printing and conversion
        to bytes.
        """
        return (
            f"origin:{self.origin},"
            f"dest:{self.dest},"
            f"src:{'.'.join(self.src)},"
            f"seqno:{self.seqno},"
            f"time:{self.time},"
            f"type:{self.type}"
        )

    def to_bytes(self):
        """Returns a byte array representing the frame for sending over the network
        """
        return self.to_string().encode("utf-8")

    def from_string(self, in_str: str):
        """Parses the string representation of a frame (that has possibly arrived
        over the network) and fills the fields with the information it contains.
        """
        fields = {}
        pairs = in_str.split(",")
        for pair in pairs:
            split_pair = pair.split(":")
            fields[split_pair[0]] = split_pair[1]

        self.origin = fields["origin"]
        self.dest = fields["dest"]
        self.src = fields["src"].split(".")
        self.seqno = int(fields["seqno"])
        self.time = int(fields["time"])
        self.type = FrameType(int(fields["type"]))

File: src/frame.cpp
/*
Implementation of the Frame class
*/
#include "frame.h"
/*
Constructor for Frame. 

Arguments:
  origin: the name of the node where the frame originated
  dest: the name of the destination node of the frame
  src: a list of strings containing the names of the nodes the frame has passed through in order
  seqno: the sequence number applied to the frame by the original sender
  time: the time, in minutes after midnight that the previous sender can arrive at the recipient
  type: Enum for frametype
*/
Frame::Frame(string origin, string dest, list<string> &src, int seqno, int time,
             FrameType type)
    : origin(origin), dest(dest), src(src), seqno(seqno), time(time),
      type(type) {}

/*
No args constructor for Frame
*/
Frame::Frame() {}

/*
Utility method to convert a Frame object to a string for easy printing

Arguments:
  None
Returns:
  String representation of the frame
*/
string Frame::to_string() {
  ostringstream src_string;
  ostringstream main_string;
  list<string>::iterator it;
  for (it = src.begin(); it != src.end(); ++it) {
    src_string << *it;
    if (it != --src.end()) {
      src_string << ".";
    }
  }
  main_string << "origin:" << origin << ","
              << "dest:" << dest << ","
              << "src:" << src_string.str() << ","
              << "seqno:" << seqno << ","
              << "time:" << time << ","
              << "type:" << type;
  return main_string.str();
}

/*
Method for initialising a Frame object from a string representation.

Used to initialise a Frame object from a string that has arrived over the network.
Iterates through the string breaking it up into tokens, and then erasing the tokens from the string.
The src is first ingested as a string and then broken up into a list.

Arguments:
  in_str: A string representation of a frame
Returns:
  void
*/

void Frame::from_string(string &in_str) {
  map<string, string> tokens;
  string delimiter = ",";
  string inner_delimiter = ":";
  string token;
  string key;
  string value;
  size_t pos = 0;
  size_t inner_pos;
  while ((pos = in_str.find(delimiter)) != string::npos) {
    inner_pos = 0;
    token = in_str.substr(0, pos);
    inner_pos = token.find(inner_delimiter);
    // We don't want the inner delimiter in our string
    key = token.substr(0, inner_pos);
    value = token.substr(inner_pos + 1, token.length());
    tokens.insert({key, value});
    in_str.erase(0, pos + delimiter.length());
  }
  // Run one last time for last token
  inner_pos = 0;
  token = in_str.substr(0, pos);
  inner_pos = token.find(inner_delimiter);
  key = token.substr(0, inner_pos);
  value = token.substr(inner_pos + 1, token.length());
  tokens.insert({key, value});
  in_str.erase(0, pos + delimiter.length());

  string src_string;
  try {
    origin = tokens.at("origin");
    dest = tokens.at("dest");
    seqno = atoi(tokens.at("seqno").c_str());
    time = atoi(tokens.at("time").c_str());
    type = static_cast<FrameType>(atoi(tokens.at("type").c_str()));
    src_string = tokens.at("src");
  } catch (const std::out_of_range &oor) {
    cout << "Parsing frame failed, exiting" << endl;
    fflush(stdout);
    exit(1);
  }
  string src_delimiter = ".";
  pos = 0;
  while ((pos = src_string.find(src_delimiter)) != string::npos) {
    token = src_string.substr(0, pos);
    src.push_back(token);
    src_string.erase(0, pos + delimiter.length());
  }
  // One last time for the final name in the string
  token = src_string.substr(0, pos);
  src.push_back(token);
  src_string.erase(0, pos + delimiter.length());
}